### ISSUE TITLE
How to configure the desired rateLimit?

### ISSUE THREAD

**Issue by mjzarrin** [ccxt/ccxt#5565]
- OS: MacOS
- Programming Language version: Python (3.7.3)
- CCXT version:  1.18.965
- Exchange: Bitmex
- Method: Bitmex.fetch_ticker(), Bitmex.sign()

It seems that authentication doesn't work on Bitmex public API. consequently [the Rate Limit is more restricted](https://www.bitmex.com/app/restAPI#Request-Rate-Limits).

Here is a test code.

    import time
    import ccxt
    from time import sleep    

    def get_epochtime_ms():
        return int(time.time() * 1000.0)    
    
    bitmex = ccxt.bitmex({
        'apiKey': 'API_KEY',
        'secret': 'SECRET',
        'timeout': 30000,
        # 'enableRateLimit': True,
    })    

    bitmex.load_markets()    

    while True:
        try:    

            fetching_timestamp = get_epochtime_ms()
            btc_usd_ticker = bitmex.fetch_ticker('BTC/USD')
            fetched_timestamp = get_epochtime_ms()
            server_timestamp = btc_usd_ticker['timestamp']    

            print('server_timestamp  ', server_timestamp)
            print('fetching_timestamp', fetching_timestamp)
            print('fetched_timestamp ', fetched_timestamp)
            print('################################')    

            sleep_time = (1000 + fetching_timestamp - get_epochtime_ms()) / 1000
            if sleep_time > 0:
                sleep(sleep_time)
        except Exception as e:
            print('Got Exception :| ', e)
            sleep(2)

**Comment by frosty00** [comment]
Hi there @mjzarrin,

Authentication does indeed work on the bitmex. This is the output I get running your script:

```
server_timestamp   1563710200772
fetching_timestamp 1563710242313
fetched_timestamp  1563710242383
################################
server_timestamp   1563710202151
fetching_timestamp 1563710243317
fetched_timestamp  1563710243380
################################
server_timestamp   1563710203231
fetching_timestamp 1563710244319
fetched_timestamp  1563710244383
################################
server_timestamp   1563710203231
fetching_timestamp 1563710245323
fetched_timestamp  1563710245387
################################
server_timestamp   1563710203231
fetching_timestamp 1563710246328
fetched_timestamp  1563710246389
################################
```

Maybe you're computers system time if out of sync, you can try to use `sudo ntpdate -u pool.ntp.org` to fix the time

**Comment by mjzarrin** [comment]
@frosty00 
Thank you for the testing. Actually, you should run the code till it has thrown Rate_Limit exception. it takes about a minute.
As I mentioned, Bitmex Rate_Limit for authenticated clients is 60 requests per min. in my test code, I would expect to fetch tickers every second. actually, every Rate_Limit more than 30 requests per minute will throw the Rate_Limit exception. Although I could not find any error or clue that shows the client authentication status, The test results show the client is most probably not authenticated.
here is a sample result after 1 minute.


     ...
     server_timestamp   1563722225081
     fetching_timestamp 1563722225727
     fetched_timestamp  1563722226010
     ################################
     server_timestamp   1563722225388
     fetching_timestamp 1563722226728
     fetched_timestamp  1563722226919
     ################################
     server_timestamp   1563722226483
     fetching_timestamp 1563722227732
     fetched_timestamp  1563722228001
     ################################
     Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
     server_timestamp   1563722230258
     fetching_timestamp 1563722230922
     fetched_timestamp  1563722231126
     ################################
     server_timestamp   1563722230997
     fetching_timestamp 1563722231923
     fetched_timestamp  1563722232178
     ################################
     Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
     server_timestamp   1563722233488
     fetching_timestamp 1563722235096
     fetched_timestamp  1563722235380
     ################################
     server_timestamp   1563722233488
     fetching_timestamp 1563722236101
     fetched_timestamp  1563722236305
     ################################
     Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
     server_timestamp   1563722239203
     fetching_timestamp 1563722239275
     fetched_timestamp  1563722239607
     ################################
     server_timestamp   1563722239656
     fetching_timestamp 1563722240278
     fetched_timestamp  1563722240692
     ################################
     Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
     server_timestamp   1563722242603
     fetching_timestamp 1563722243470
     fetched_timestamp  1563722243662
     ################################
     server_timestamp   1563722243727
     fetching_timestamp 1563722244475
     fetched_timestamp  1563722244725
     ################################
     Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
     server_timestamp   1563722245728
     fetching_timestamp 1563722247647
     fetched_timestamp  1563722247885
     ################################


Also, I've checked [Bitmex Sign function](https://github.com/ccxt/ccxt/blob/bb2adda1e6a890e94cdc5c9dc5798679d7eaa40b/python/ccxt/bitmex.py#L1155) by debugging. It seems for public API calls (like fetch_ticker), the method skips the auth process which makes the problem.

**Comment by frosty00** [comment]
There is no bug here, remove the override of `timeout` in the options and enable `enableRateLimit` and the code will run indefinitely. @kroitor pls close.

**Comment by mjzarrin** [comment]
@frosty00 
Close?!?!? 

Certainly, `timeout` doesn't have any roles in this issue. 
Here is your requested script.  
    
    import ccxt
    import time
    import pandas as pd    
   
    def get_epochtime_ms():
        return int(time.time() * 1000.0)    

    bitmex = ccxt.bitmex({
        'apiKey': 'API_KEY',
        'secret': 'SECRET',
        'enableRateLimit': True,
    })    

    bitmex.load_markets()    

    i = 0
    while True:
        try:
            btc_usd_ticker = bitmex.fetch_ticker('BTC/USD')
            print(i, pd.to_datetime(get_epochtime_ms(),unit='ms'))
            i += 1
        except Exception as e:
            print('Got Exception :| ', e)

By providing API_KEY and SECRET, I would expect the rate_limit increases from 30 Requests Per Min(RPM) for non_authenticated clients to 60 RPM for authenticated clients. for more info refer [here](https://www.bitmex.com/app/restAPI#Request-Rate-Limits) please.
However, the result shows enabling `enableRateLimit ` would apply 2 seconds delay(= 30 RPM) which is unacceptable for authenticated client.

    0 2019-07-21 21:41:32.362000
    1 2019-07-21 21:41:34.446000
    2 2019-07-21 21:41:36.551000
    3 2019-07-21 21:41:38.392000
    4 2019-07-21 21:41:40.405000
    5 2019-07-21 21:41:42.395000
    6 2019-07-21 21:41:44.565000
    7 2019-07-21 21:41:46.594000
    8 2019-07-21 21:41:48.389000
    9 2019-07-21 21:41:50.551000
    10 2019-07-21 21:41:52.710000
    11 2019-07-21 21:41:54.574000
    12 2019-07-21 21:41:56.603000
    13 2019-07-21 21:41:58.529000
    14 2019-07-21 21:42:00.461000
    15 2019-07-21 21:42:02.486000
    16 2019-07-21 21:42:04.485000
    17 2019-07-21 21:42:06.410000
    18 2019-07-21 21:42:08.470000
    19 2019-07-21 21:42:10.550000
    20 2019-07-21 21:42:12.486000
    21 2019-07-21 21:42:14.529000
    22 2019-07-21 21:42:16.591000
    23 2019-07-21 21:42:18.632000
    24 2019-07-21 21:42:20.520000
    25 2019-07-21 21:42:22.431000
    26 2019-07-21 21:42:24.423000
    27 2019-07-21 21:42:26.449000
    28 2019-07-21 21:42:28.511000
    29 2019-07-21 21:42:31.418000
    30 2019-07-21 21:42:32.852000

Any thoughts?
@kroitor

**Comment by kroitor** [comment]
@mjzarrin you're supposed to tweak the `rateLimit` according to your need, as described in the Manual here: https://github.com/ccxt/ccxt/wiki/Manual#rate-limit.

> The exchange.rateLimit property is set to a safe default which is sub-optimal. Some exchanges may have varying rate limits for different endpoints. It is up to the user to tweak rateLimit according to application-specific purposes.

This was answered here:

- https://github.com/ccxt/ccxt/issues/1855#issuecomment-364623222
- https://github.com/ccxt/ccxt/issues/4779#issuecomment-469249844
- https://github.com/ccxt/ccxt/issues/4600#issuecomment-459741049
- https://github.com/ccxt/ccxt/issues/2700#issuecomment-385483846
- https://github.com/ccxt/ccxt/issues/2041
- https://github.com/ccxt/ccxt/issues/4605#issuecomment-460013736

Let us know if the above does not help or does not answer the question.

**Comment by mjzarrin** [comment]
@kroitor, 
First of all, I appreciate your great works on developing CCXT and keeping it update.
Regarding the problem, I read all materials that you provide, However none of them is exactly like this issue.
Here is [Bitmex simple `rateLimit` rule](https://www.bitmex.com/app/restAPI#Limits). 

> Requests to our REST API are rate limited to 60 requests per 1 minute. This counter refills continuously. If you are not logged in, your `rateLimit` is 30/1 minute.

The problem is, although the client is authenticated, it can not reach to it's desired quota. Please check the following script and its output.

    import time
    import ccxt
    import pandas as pd    

    def get_epochtime_ms():
        return int(time.time() * 1000.0)    

    bitmex = ccxt.bitmex({
        'apiKey': 'FILL_YOUR_API_KEY',
        'secret': 'FILL_YOUR_SECRET',
        'enableRateLimit': True,
        'rate_limit': 1100 # (ms)
    })    

    bitmex.load_markets()    

    i = 0
    while True:
        try:
            btc_usd_ticker = bitmex.fetch_ticker('BTC/USD')
            print(i, pd.to_datetime(get_epochtime_ms(), unit='ms'))
            i += 1
        except Exception as e:
            print('Got Exception :| ', e)

and here is the output. after 30 requests Bitmex sends rateLimit exception. 

    0 2019-07-22 10:15:14.298000
    1 2019-07-22 10:15:15.422000
    2 2019-07-22 10:15:16.480000
    3 2019-07-22 10:15:17.585000
    4 2019-07-22 10:15:18.683000
    5 2019-07-22 10:15:19.795000
    6 2019-07-22 10:15:20.888000
    7 2019-07-22 10:15:21.994000
    8 2019-07-22 10:15:23.125000
    9 2019-07-22 10:15:24.197000
    10 2019-07-22 10:15:25.332000
    11 2019-07-22 10:15:26.443000
    12 2019-07-22 10:15:27.491000
    13 2019-07-22 10:15:28.592000
    14 2019-07-22 10:15:29.733000
    15 2019-07-22 10:15:30.843000
    16 2019-07-22 10:15:31.906000
    17 2019-07-22 10:15:33.642000
    18 2019-07-22 10:15:34.105000
    19 2019-07-22 10:15:35.234000
    20 2019-07-22 10:15:36.296000
    21 2019-07-22 10:15:37.441000
    22 2019-07-22 10:15:38.506000
    23 2019-07-22 10:15:39.600000
    24 2019-07-22 10:15:40.696000
    25 2019-07-22 10:15:41.799000
    26 2019-07-22 10:15:42.901000
    27 2019-07-22 10:15:43.999000
    28 2019-07-22 10:15:45.156000
    29 2019-07-22 10:15:46.241000
    30 2019-07-22 10:15:47.383000
    Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
    31 2019-07-22 10:15:49.547000
    Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
    32 2019-07-22 10:15:51.713000
    33 2019-07-22 10:15:52.814000
    Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
    34 2019-07-22 10:15:55.012000
    Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
    35 2019-07-22 10:15:57.224000
    Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
    36 2019-07-22 10:15:59.420000
    Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
    37 2019-07-22 10:16:01.618000
    38 2019-07-22 10:16:03.685000
    ...

Even with `'rate_limit': 1500 # (ms)` , this output, has rateLimit exception too. 

    ...
    110 2019-07-22 10:23:29.282000
    111 2019-07-22 10:23:30.843000
    112 2019-07-22 10:23:32.384000
    113 2019-07-22 10:23:33.839000
    114 2019-07-22 10:23:35.377000
    115 2019-07-22 10:23:36.893000
    Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
    116 2019-07-22 10:23:39.844000
    117 2019-07-22 10:23:41.346000
    118 2019-07-22 10:23:42.947000
    Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
    119 2019-07-22 10:23:45.856000
    120 2019-07-22 10:23:47.362000
    121 2019-07-22 10:23:48.865000
    Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
    122 2019-07-22 10:23:51.882000
    123 2019-07-22 10:23:53.356000
    124 2019-07-22 10:23:54.976000
    Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
    125 2019-07-22 10:23:57.858000
    126 2019-07-22 10:23:59.409000
    127 2019-07-22 10:24:00.999000
    Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
    128 2019-07-22 10:24:03.912000
    129 2019-07-22 10:24:05.406000
    130 2019-07-22 10:24:06.907000
    Got Exception :|  bitmex {"error":{"message":"Rate limit exceeded, retry in 1 seconds.","name":"RateLimitError"}}
    ...

Indeed, Bitmex doesn't increase rateLimit for authenticated clients. What do you think about the cause of this restriction?

**Comment by kroitor** [comment]
> What do you think about the cause of this restriction?

The actual reason is that CCXT does not sign public requests for BitMEX, so it didn't work for you with the public `fetchTicker` call. That's an easy fix though. 

Actually, very few exchanges will require authentication (signing) for public requests. And BitMEX `fetchTicker` is a public endpoint: https://www.bitmex.com/api/v1/instrument/activeAndIndices. 

I tried to replicate your error (with public requests unsigned), but in my case it did run for 1 minute or so before breaking down. Looks like their rate limits may vary depending on your location.

Then I tried with signed public calls, and it did work (apparently) – so it continued running without breaking down. Tested it in Python with your snippet and in JS with `node examples/js/cli bitmex fetchTicker BTC/USD --poll --verbose` (with `rateLimit` set to 1100).

I've uploaded the fix, and it will sign the public requests starting from CCXT 1.18.973 (make sure to update your version), if you supply the `apiKey` and `secret` as you do.

Let us know if that does not help. We will be happy if you report back whether it does the trick for you or not. Thx for your involvement!
